msticnb/nb/azsent/alert/ti_enrich.py

```diff
@@ -0,0 +1,333 @@
+
+        Raises
+        ------
+        MsticnbMissingParameterError
+            If required parameters are missing
+
+        MsticnbDataProviderError
+            If data is not avaliable
+
+        """
+        # This line use logic in the superclass to populate options
+        # (including default options) into this class.
+        super().run(
+            value=value, data=data, timespan=timespan, options=options, **kwargs
+        )
+
+        if not timespan and data is None:
+            raise MsticnbMissingParameterError("timespan.")
+
+        # If data is not provided, query Sentinel to get it.
+        if data is None:
+            nb_print("Collecting alerts")
+            if value is not None:
+                data = _get_all_alerts(self.query_provider, timespan, value)
+            else:
+                data = _get_all_alerts(self.query_provider, timespan)
+
+        # Create a result class
+        # Add description to results for context
+        self._last_result = TIEnrichResult(
+            description=f"""Enriched alerts,
+                            with the filter of {value}"""
+        )
+
+        # Establish TI providers
+        if "tilookup" in self.data_providers.providers:
+            ti_prov = self.data_providers.providers["tilookup"]
+        else:
+            raise MsticnbDataProviderError("No TI providers avaliable")
+
+        if isinstance(data, pd.DataFrame) and not data.empty:
+            data["Entities"] = data["Entities"].apply(_entity_load)
+            tqdm.pandas(desc="TI lookup progress")
+            ti_sec = False
+            if "secondary" in self.options:
+                ti_sec = True
+            md(
+                """Alerts enriched with threat intelligence -
+                 TI Risk is the the hightest score provided by any of the configured providers."""
+            )
+            data["TI Risk"] = data.progress_apply(
+                lambda row: _lookup(row, ti_prov, secondary=ti_sec), axis=1
+            )
+            display(
+                data[
+                    ["StartTimeUtc", "AlertName", "Severity", "TI Risk", "Description"]
+                ]
+                .sort_values(by=["StartTimeUtc"])
+                .style.applymap(_color_cells)
+                .hide_index()
+            )
+            if "details" in self.options:
+                alert_pick = _alert_picker(data, ti_prov, secondary=ti_sec)
+        else:
+            raise MsticnbDataProviderError("No alerts avaliable")
+
+        self._last_result.enriched_results = data
+        self._last_result.picker = alert_pick
+
+        return self._last_result
+
+
+# %%
+# Display Alert Picker
+@set_text(
+    title="Select Alert",
+    text="""
+Select and alert to view further details.
+""",
+)
+def _alert_picker(data, ti_prov, secondary):
+    ti_provs = "primary"
+    if secondary is True:
+        ti_provs = "all"
+
+    def show_full_alert(alert):
```

Comment:
We'll have to revisit this in the light of the discussions with the notebooks team about the use of the Output widget.
I think the display action functions will have to do something like return a list of displayable objects rather than do the displaying themselves